### PR TITLE
Wasm implementation of chacha20poly1305

### PR DESCRIPTION
**Pull request by tuyennhv**
**Motivation**

We want to improve `chacha20poly1305` in lodestar, see https://github.com/ChainSafe/lodestar/issues/4652

**Description**
- Wasm implementation of chacha20, poly1305 and chacha20poly1305
- Some optimization compared to stablelib:
  - No memory allocation inside the assemblyscript implementation but we have to do the chunkify (i.e. multiple update() calls)
  - DATA_CHUNK_LENGTH is set to 65536 (the same to libp2p-noise), so it takes only 1 `update()` call if consumed from libp2p-noise
  - Support `overwriteSealed` which to leverage the input when decrypting (`open` function), which save a memory allocation

part of https://github.com/ChainSafe/lodestar/issues/4652

**Benchmark result (commit 26d2441)**

|Data size|Method|Implementation|Speed (time per op)|
|---------|--------|---------------|-------|
|512|seal|this|8.41 ns|
|512|seal|stablelib|12.13 ns|
|512|open|this|7.73 ns|
|512|open|stablelib|12.42 ns|
|4096|seal|this|52.44 ns|
|4096|seal|stablelib|62.38 ns|
|4096|open|this|47.7 ns|
|4096|open|stablelib|72.39 ns|
|16384|seal|this|201.58 ns|
|16384|seal|stablelib|273.92 ns|
|16384|open|this|201.24 ns|
|16384|open|stablelib|318.42 ns|
|65536|seal|this|836.59 ns|
|65536|seal|stablelib|1.19 us|
|65536|open|this|802.29 ns|
|65536|open|stablelib|1.35 us|

### CODE REVIEW

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+// Applies the ChaCha core function to 16-byte input,
+// 32-byte key key, and puts the result into 64-byte array out.
+function core(out: usize, input: usize, key: usize): void {
+  const j0: i32 = 1634760805; // 0x61707865; // "expa"  -- ChaCha's "sigma" constant
+  const j1: i32 = 857760878; // 0x3320646e; // "nd 3"     for 32-byte keys
+  const j2: i32 = 2036477234; // 0x79622d32; // "2-by"
+  const j3: i32 = 1797285236; // 0x6b206574; // "te k"
+  const j4: i32 = (i32(load8(key, 3)) << 24) | (i32(load8(key, 2)) << 16) | (i32(load8(key, 1)) << 8) | i32(load8(key, 0));
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:98]:
```suggestion
  const j4: u32 = bswap(load32(key));
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  let x11: i32 = j11;
+  let x12: i32 = j12;
+  let x13: i32 = j13;
+  let x14: i32 = j14;
+  let x15: i32 = j15;
+
+  for (let i: i32 = 0; i < ROUNDS; i += 2) {
+    x0 = (x0 + x4) | 0;
+    x12 ^= x0;
+    x12 = (x12 >>> (32 - 16)) | (x12 << 16);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:131]:
```suggestion
    x12 = rotl(x12, 16)
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    x8 = (x8 + x12) | 0;
+    x4 ^= x8;
+    x4 = (x4 >>> (32 - 12)) | (x4 << 12);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:134]:
```suggestion
    x4 = rotl(x4, 12);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  let x12: i32 = j12;
+  let x13: i32 = j13;
+  let x14: i32 = j14;
+  let x15: i32 = j15;
+
+  for (let i: i32 = 0; i < ROUNDS; i += 2) {
+    x0 = (x0 + x4) | 0;
+    x12 ^= x0;
+    x12 = (x12 >>> (32 - 16)) | (x12 << 16);
+    x8 = (x8 + x12) | 0;
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:132]:
```suggestion
    x8 = x8 + x12;
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+    x4 ^= x8;
+    x4 = (x4 >>> (32 - 12)) | (x4 << 12);
+    x1 = (x1 + x5) | 0;
+    x13 ^= x1;
+    x13 = (x13 >>> (32 - 16)) | (x13 << 16);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:137]:
```suggestion
    x13 = rotl(x13, 16);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    x3 = (x3 + x7) | 0;
+    x15 ^= x3;
+    x15 = (x15 >>> (32 - 16)) | (x15 << 16);
+    x11 = (x11 + x15) | 0;
+    x7 ^= x11;
+    x7 = (x7 >>> (32 - 12)) | (x7 << 12);
+
+    x2 = (x2 + x6) | 0;
+    x14 ^= x2;
+    x14 = (x14 >>> (32 - 8)) | (x14 << 8);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:157]:
```suggestion
    x14 = rotl(x14, 8);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    x12 = (x12 >>> (32 - 16)) | (x12 << 16);
+    x8 = (x8 + x12) | 0;
+    x4 ^= x8;
+    x4 = (x4 >>> (32 - 12)) | (x4 << 12);
+    x1 = (x1 + x5) | 0;
+    x13 ^= x1;
+    x13 = (x13 >>> (32 - 16)) | (x13 << 16);
+    x9 = (x9 + x13) | 0;
+    x5 ^= x9;
+    x5 = (x5 >>> (32 - 12)) | (x5 << 12);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:140]:
```suggestion
    x5 = rotl(x5, 12);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+    x2 = (x2 + x6) | 0;
+    x14 ^= x2;
+    x14 = (x14 >>> (32 - 16)) | (x14 << 16);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:144]:
```suggestion
    x14 = rotl(x14, 16);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    x10 = (x10 + x14) | 0;
+    x6 ^= x10;
+    x6 = (x6 >>> (32 - 12)) | (x6 << 12);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:147]:
```suggestion
    x6 = rotl(x6, 12);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    x3 = (x3 + x7) | 0;
+    x15 ^= x3;
+    x15 = (x15 >>> (32 - 16)) | (x15 << 16);
+    x11 = (x11 + x15) | 0;
+    x7 ^= x11;
+    x7 = (x7 >>> (32 - 12)) | (x7 << 12);
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:153]:
```suggestion
    x7 = rotl(x7, 12);
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts (hunk 6 above)
Comment (MaxGraey) [reply packages/as-chacha20poly1305/assembly/chacha20.ts:157]: and etc

File: packages/as-chacha20poly1305/assembly/util.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  }
+}
+
+export function wipe16(array: usize, length: u32): void {
+  // Right now it's similar to array.fill(0). If it turns
+  // out that runtimes optimize this call away, maybe
+  // we can try something else.
+  for (let i: u32 = 0; i < length; i++) {
+    store16(array, i, 0);
+  }
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/util.ts:60]:
```suggestion
  memory.fill(array, 0, length * sizeof<u16>())
```

File: packages/as-chacha20poly1305/assembly/chacha20.ts [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+  let x8: i32 = j8;
+  let x9: i32 = j9;
+  let x10: i32 = j10;
+  let x11: i32 = j11;
+  let x12: i32 = j12;
+  let x13: i32 = j13;
+  let x14: i32 = j14;
+  let x15: i32 = j15;
+
+  for (let i: i32 = 0; i < ROUNDS; i += 2) {
```
Comment (MaxGraey) [inline packages/as-chacha20poly1305/assembly/chacha20.ts:128]:
```suggestion
  for (let i = 0; i < ROUNDS; i += 2) {
```

**Comment by dapplion** [comment]
@tuyennhv chacha20poly is unrelated to SSZ. I think it would make more sense for this implementation to be in a standalone repo with the package name instead of mixed in here.

**Comment by wemeetagain** [comment]
Agree with @dapplion 

Can you PR this into https://github.com/ChainSafe/as-chacha20poly1305

**Comment by tuyennhv** [comment]
thanks @MaxGraey so much for all of your comments, it's so helpful to me as I'm new to assemblyscript. I'll revise in this PR https://github.com/ChainSafe/as-chacha20poly1305/pull/1

**Pull request closed without merging**